### PR TITLE
feat(material/select): allow user-defined aria-describedby

### LINKED ISSUE #16209: Mat-hint causing an ExpressionChangedAfterItHasBeenCheckedError if respective input has a *ngIf directive
#### Reproduction

[Minimal repro](https://stackblitz.com/edit/angular-x3assq?file=app%2Fform-field-overview-example.html
) on StackBlitz. Open the console to see the following error message:

```
Error: ExpressionChangedAfterItHasBeenCheckedError: Expression has changed after it was checked. Previous value: 'aria-describedby: null'. Current value: 'aria-describedby: mat-hint-0'.
```
This always happens if the respective input has a `*ngIf`-directive. I couldn't observe any subsequent errors caused by this. Using the `hintLabel` property instead of a `<mat-hint>` results in the same error. 
 
#### Expected Behavior

No `ExpressionChangedAfterItHasBeenCheckedError` when using *ngIf on an input

#### Actual Behavior

`ExpressionChangedAfterItHasBeenCheckedError`

#### Environment

- Angular: `^7.1.4`
- CDK/Material: `^7.2.0`
- Browser(s): Chrome (Most recent @ Version 75.0.3770.80)
- Operating System (e.g. Windows, macOS, Ubuntu): macOS

### PR DESCRIPTION
**Pull request by RobertAKARobin**
Based on #24292.

Further fixes #16209.

### CODE REVIEW

**Comment by RobertAKARobin** [comment]
I suspect `ci/circleci: e2e_tests` just needs to be rerun.

File: src/material/select/select.ts [hunk 1]
```diff
@@ -611,7 +614,7 @@ export abstract class _MatSelectBase<C>
   ngOnChanges(changes: SimpleChanges) {
     // Updating the disabled state is handled by `mixinDisabled`, but we need to additionally let
     // the parent form field know to run change detection when the disabled state changes.
-    if (changes['disabled']) {
+    if (changes['disabled'] || changes['userAriaDescribedBy']) {
```
Comment (crisbeto) [inline src/material/select/select.ts:617]: Why is this change necessary? Since the `aria-describedby` is on the select it shouldn't be required.

File: src/material-experimental/mdc-select/select.spec.ts [hunk 2]
```diff
@@ -222,6 +222,22 @@ describe('MDC-based MatSelect', () => {
           expect(select.getAttribute('tabindex')).toEqual('0');
         }));
 
+        it('should set `aria-describedby` to the id of the mat-hint', fakeAsync(() => {
+          expect(select.getAttribute('aria-describedby')).toBeNull();
+
+          fixture.componentInstance.hint = 'test';
+          fixture.detectChanges();
+          const hint = fixture.debugElement.query(By.css('mat-hint')).nativeElement;
+          expect(select.getAttribute('aria-describedby')).toBe(hint.getAttribute('id'));
```
Comment (crisbeto) [inline src/material-experimental/mdc-select/select.spec.ts:231]: I think that you can drop this assertion since it's covered by the one below.

File: src/material/select/select.ts (hunk 1 above)
Comment (RobertAKARobin) [reply src/material/select/select.ts:617]: Without it, this test fails:

```
it('should support user binding to `aria-describedby`', fakeAsync(() => {
  fixture.componentInstance.ariaDescribedBy = 'test';
  fixture.detectChanges();
  expect(select.getAttribute('aria-describedby')).toBe('test');
}));
```

As a result of this PR (and #24292 for MatChipList and #19587 for MatInput), the `aria-describedby` is no longer set in `@Component({ host })`. Rather, it is set using `nativeElement.setAttribute` in `MatSelect#setDescribedByIds`, which is called by `MatFormField#_syncDescribedByIds`, which in turn is called when `stateChanges` updates.

So, `userAriaDescribedBy` has to trigger a `stateChanges.next()` in order for the `aria-describedby` to actually be set.

File: src/material-experimental/mdc-select/select.spec.ts (hunk 2 above)
Comment (RobertAKARobin) [reply src/material-experimental/mdc-select/select.spec.ts:231]: This asserts that it's exactly the same as the hint's ID, whereas the next assertion asserts only that it matches a certain format: `expect(select.getAttribute('aria-describedby')).toMatch(/^mat-mdc-hint-\d+$/);`

File: src/material/select/select.ts (hunk 1 above)
Comment (RobertAKARobin) [reply src/material/select/select.ts:617]: ...and for that matter, MatChipList needs unit tests for this too! I've submitted #24657. In that one I used a setter for `userAriaDescribedBy` to call `stateChanges.next()`. Let me know if you'd prefer that here too.

**Review by crisbeto — APPROVED** [review]
LGTM
